### ISSUE TITLE
Foundation PTv3 encoders: Sonata and Utonia license, input contract, loadability

### ISSUE THREAD

**Issue by Chitchai-Jantanarak** [Chitchai-Jantanarak/PTv3-building-segment#26]
## Question

For Sonata and Utonia (pretrained PTv3 encoders): what are the exact weight licenses, the input feature contract (coordinates only, or normals / color / intensity required), the grid size and channel layout they were trained with, and can either load into this repo's PTv3 (`configs/model/ptv3.yaml`: `in_channels: 4`, `enc_channels: [32, 64, 128, 256, 512]`, `enc_depths: [2, 2, 2, 6, 2]`) without architectural changes?

## Why this blocks decisions

The representation baseline compares our MAE against a foundation encoder and random init on the same frozen binary probe. That comparison is only fair if the foundation encoder can be loaded and fed the features it expects. The CC-BY-NC question decides whether a foundation encoder is a benchmark control only, or a candidate for the production pipeline.

## Resolution

A comment listing, per encoder: license (with URL), input contract, native grid size, architecture deltas from this repo's PTv3, checkpoint URL and size, and a yes / no on loadability with the concrete change needed if no.

**Comment by Chitchai-Jantanarak** [comment]
# Issue #26 research: Sonata / Utonia loadability against this repo's PTv3

Repo config for reference (`configs/model/ptv3.yaml`): `in_channels=4` (xyz + rel_z; RGB goes to a separate color head), `enc_channels=[32,64,128,256,512]`, `enc_depths=[2,2,2,6,2]`, `dec_channels=[64,64,128,256]`, `dec_depths=[2,2,2,2]`, `patch_size=512`, `grid_size=0.04`.

---

## 1. Sonata

### License
- **Code**: Apache License 2.0, Meta / facebookresearch. Repo: [facebookresearch/sonata](https://github.com/facebookresearch/sonata), license file at [sonata/LICENSE](https://github.com/facebookresearch/sonata/blob/main/LICENSE).
- **Weights**: **CC-BY-NC 4.0** (non-commercial). README states verbatim: *"Sonata weight is released under the CC-BY-NC 4.0 license (restricted by NC of datasets like HM3D, ArkitScenes)"* — [facebookresearch/sonata README](https://github.com/facebookresearch/sonata).

### Input contract
- Required fields per point cloud: `coord` (N,3), `color` (N,3), `normal` (N,3). Optional: `segment`, `batch`.
- Confirmed via the actual model config on Hugging Face ([facebook/sonata/config.json](https://huggingface.co/facebook/sonata/blob/main/config.json)): **`in_channels: 9`** — consistent with coord(3)+color(3)+normal(3) concatenated into the feature tensor.
- Normals are **not** computed internally by the released inference code; the README's preprocessing pipeline expects them as an input field. Whether Sonata degrades gracefully with a zero-filled normal/color (as Utonia explicitly documents) is **not found** in the README — no fallback behavior is stated for Sonata specifically.
- Default preprocessing grid size for inference: `grid_size=0.02`. Source: [github.com/facebookresearch/sonata README](https://github.com/facebookresearch/sonata).

### Grid size / training domain
- Grid size: `0.02` (indoor point-cloud scale).
- Training datasets, all **indoor**: ScanNet, S3DIS, Structured3D, ArkitScenes, HM3D. Source: repo README and project page ([projectaria.com/news/introducing-sonata](https://www.projectaria.com/news/introducing-sonata/)).
- No outdoor, ALS, or MLS training domain found anywhere in the repo, paper, or project page.

### Architecture
From the real `config.json` at [huggingface.co/facebook/sonata](https://huggingface.co/facebook/sonata/blob/main/config.json):

```
in_channels:      9
enc_channels:     [48, 96, 192, 384, 512]
enc_depths:       [3, 3, 3, 12, 3]
enc_num_head:     [3, 6, 12, 24, 32]
enc_patch_size:   [1024, 1024, 1024, 1024, 1024]
dec_channels:     [96, 96, 192, 384]
dec_depths:       [3, 3, 3, 3]
dec_num_head:     [6, 6, 12, 32]
stride:           [2, 2, 2, 2]
order:            [z, z-trans, hilbert, hilbert-trans]
enable_flash:     true
mask_token:       true   (SSL pretext-specific)
enc_mode:         true
```

- Deepest-stage embedding dim: **512**.
- The decoder shown here is the lightweight per-point decoder used for the masked-point / self-distillation SSL objective during pretraining, not a segmentation decoder. Whether the released "downstream-ready" checkpoint ships these decoder weights or is encoder-only depends on which file you load (see below).
- Note a minor inconsistency between two source fetches on `dec_patch_size` (4 vs 5 entries reported); not load-bearing for the loadability verdict.

### Checkpoints
On Hugging Face, [facebook/sonata](https://huggingface.co/facebook/sonata):
| file | size | contents |
|---|---|---|
| `pretrain-sonata-v1m1-0-base.pth` | 1.95 GB | full pretraining checkpoint (encoder + SSL decoder/EMA state) |
| `sonata.pth` / `model.safetensors` | 434 MB each | encoder-only checkpoint for downstream use |
| `sonata_small.pth` | 155 MB | smaller encoder variant |
| `sonata_linear_prob_head_sc.pth` | 100 kB | a pre-fit linear probe head (ScanNet) |

Load via Hugging Face hub:
```python
sonata.model.load("sonata", repo_id="facebook/sonata")
# or
PointTransformerV3.from_pretrained("facebook/sonata")
```
Cached at `~/.cache/sonata/ckpt`. Source: [facebookresearch/sonata README](https://github.com/facebookresearch/sonata).

### Loadability verdict: **NO** (not drop-in against the repo's existing PTv3 shape)
Weight shapes must match exactly to load a `state_dict`, so this is not a config tweak:
- `in_channels` 9 vs repo's 4 — repo would need to feed coord+color+normal as one feature tensor instead of routing color to a separate head and using rel_z instead of normal. **Normals must be computed at load time** (this repo currently doesn't compute normals anywhere).
- `enc_channels`/`enc_depths` differ at every stage except the last width (512 vs 512): `[48,96,192,384,512]` vs repo's `[32,64,128,256,512]`, depths `[3,3,3,12,3]` vs `[2,2,2,6,2]`.
- `patch_size` 1024 vs repo's 512.
- `grid_size` 0.02 vs repo's 0.04.

Practical path: instantiate a **separate** PTv3 encoder (same vendored Pointcept class, different constructor args matching Sonata's checkpoint exactly), load the pretrained weights into that standalone instance, freeze it, and feed it 9-channel input at `grid_size=0.02` for the linear probe — this does not require modifying the repo's own trained backbone, but it also is not "the repo's PTv3 with a swapped checkpoint."

### Outdoor / ALS
None. Not found anywhere in the repo, paper, or project page.

---

## 2. Utonia

### License
- **Code**: Apache License 2.0. README states: *"Utonia code is based on Sonata code, which is released by Meta under the Apache 2.0 license"* — [Pointcept/Utonia](https://github.com/Pointcept/Utonia), license file confirms Apache-2.0 SPDX.
- **Weights**: **CC-BY-NC 4.0** (non-commercial). README states verbatim: *"Utonia weight is released under the CC-BY-NC 4.0 license (restricted by NC of datasets like HM3D, ArkitScenes)"*. The Hugging Face model card license tag also reads `cc-by-nc-4.0` — [huggingface.co/Pointcept/Utonia](https://huggingface.co/Pointcept/Utonia).
- **Caveat found and resolved**: the arXiv abstract page for the Utonia paper ([arxiv.org/abs/2603.03283](https://arxiv.org/abs/2603.03283)) shows a "License: CC BY 4.0" badge. That is arXiv's license for the **paper submission text**, not the model weights — the GitHub README and the HF model-card metadata both independently and explicitly state CC-BY-NC 4.0 for the weights. Treat CC-BY-NC 4.0 as the operative weight license.

### Input contract
- Same three fields as Sonata: `coord` (N,3), `color` (N,3), `normal` (N,3), optional `segment`/`batch`.
- Explicit fallback documented (unlike Sonata): *"When color or normal is absent, please apply default zeros to the missing modality"* — [Pointcept/Utonia README](https://github.com/Pointcept/Utonia).
- Encoder-only: README states *"Utonia is a pre-trained encoder-only PTv3"* — no decoder weights ship.
- `in_channels` was **not** independently confirmed via a config file — the HF repo's `config.json` returned HTTP 404 at the expected path. Given the identical documented input contract and the explicit "code is based on Sonata code" statement, `in_channels=9` (coord+color+normal) is a strong **inference**, not a directly confirmed fact. Flagging as inferred.

### Grid size / training domain
- No single fixed grid size is published; README default example uses `grid_size=0.01` and states scale is task-dependent and should be tuned.
- Domain-specific preprocessing guidance given (this is the one thing Sonata's docs don't have): for outdoor LiDAR, disable `CenterShift` and ensure the ego-vehicle sits at the origin with the road plane aligned to the xy-plane; for single objects, enable `normalize_coord=True`.
- Training domains, per the arXiv HTML full text ([arxiv.org/html/2603.03283v1](https://arxiv.org/html/2603.03283v1)):
  - **Indoor**: ScanNet, ScanNet++, S3DIS, ARKitScenes, HM3D, Structured3D
  - **Outdoor**: nuScenes, Waymo, SemanticKITTI, and a dataset listed as "HK Remote" (name suggests a Hong Kong remote-sensing/aerial source, but this was **not confirmed** to be airborne/ALS specifically in the accessible text — flagging as unconfirmed)
  - **Object-centric**: PartNet, ScanObjectNN, GraspNet, Cap3D
  - Reported total scale: "197,868 + 1M" training samples across the mixed domains.
- This cross-domain outdoor-LiDAR training (nuScenes/Waymo/SemanticKITTI) is the single biggest difference from Sonata, which is 100% indoor.

### Architecture
Two published variants, sourced from the HF model card and arXiv text ([huggingface.co/Pointcept/Utonia](https://huggingface.co/Pointcept/Utonia), [arxiv.org/html/2603.03283v1](https://arxiv.org/html/2603.03283v1)):

| variant | params | enc_channels | enc_depths | enc_num_head |
|---|---|---|---|---|
| small | 38M | `(36, 72, 144, 252, 504)` | `(2, 2, 2, 6, 2)` | not stated |
| base | 137M | `(54, 108, 216, 432, 576)` | `(3, 3, 3, 12, 3)` | `(3, 6, 12, 24, 32)` |

- Patch size and exact decoder info: **not found** — no accessible config file confirmed `enc_patch_size`; by inheritance from the shared Sonata-derived codebase, `1024` is likely but **unconfirmed** for Utonia specifically.
- Deepest-stage embedding dim: **504** (38M variant) or **576** (137M variant).
- No decoder (encoder-only, confirmed by README).
- Interesting overlap: the 38M variant's `enc_depths` of `(2,2,2,6,2)` is an **exact match** to this repo's own `enc_depths=[2,2,2,6,2]`. Channel widths are close in shape but not identical (`36,72,144,252,504` vs repo's `32,64,128,256,512`).

### Checkpoints
On Hugging Face, [Pointcept/Utonia](https://huggingface.co/Pointcept/Utonia):
| file | size | contents |
|---|---|---|
| `pretrain-utonia-v1m1-0-base_stagev2.pth` | 7.01 GB | full pretraining state |
| `utonia.pth` | 549 MB | inference/simplified encoder checkpoint |
| `utonia_linear_prob_head_sc.pth` | 113 kB | pre-fit linear probe head (ScanNet) |

Load via:
```python
utonia.model.load("utonia", repo_id="Pointcept/Utonia").cuda()
```
Cached at `~/.cache/utonia/ckpt`.

### Loadability verdict: **NO** (not drop-in), same category as Sonata but closer domain fit
- `in_channels` almost certainly 9 vs repo's 4 (inferred, see caveat above) — same normals-must-be-computed, color-must-be-fused-not-separate-head problem as Sonata.
- `enc_channels`/`enc_depths` differ from the repo's shape at every stage except the 38M variant's depths, which match exactly. Channel widths still differ (`36,72,144,252,504` vs `32,64,128,256,512`).
- Patch size unconfirmed but presumed `1024` vs repo's `512`.
- Grid size is explicitly tunable per Utonia's own docs (`0.01` default, "fine-tune for your task") — closer in spirit to this repo's `0.04` than Sonata's fixed `0.02`, and Utonia additionally documents the outdoor-specific preprocessing (no `CenterShift`, ego-origin alignment) this repo's ALS/MLS data would actually need, which Sonata's docs don't address at all.

Same practical path as Sonata: instantiate a standalone PTv3 encoder shaped to match the checkpoint (`36,72,144,252,504` / `(2,2,2,6,2)` for the 38M variant, or the 137M shape), load Utonia's weights into it, freeze it, feed 9-channel input, and use it purely as the frozen feature extractor for the probe.

### Outdoor / ALS
- Outdoor LiDAR is one of the five explicit training domains (nuScenes, Waymo, SemanticKITTI), with dedicated outdoor preprocessing guidance in the README.
- No dedicated airborne/ALS or Thai-LiDAR example or benchmark was found. The "HK Remote" dataset name suggests a remote-sensing/aerial-adjacent source but this is **not confirmed** as ALS in the accessible text.

---

## Recommendation for the probe

Try **Utonia first**. Its pretraining explicitly spans outdoor LiDAR (nuScenes, Waymo, SemanticKITTI) in addition to indoor scenes, which is a much closer domain match to Thai airborne/mobile LiDAR than Sonata's training set, which is entirely indoor (ScanNet, S3DIS, Structured3D, ArkitScenes, HM3D). Utonia's docs also give explicit outdoor preprocessing steps (disable `CenterShift`, align the road plane to the origin) that this project's ALS/MLS pipeline would need to replicate — Sonata's docs have no equivalent guidance and no evidence of ever seeing outdoor data.

Neither loads into the repo's existing PTv3 module without an architecture swap (both need a 9-channel coord+color+normal input, computed normals, and non-matching encoder channel/depth/patch-size shapes); both would run as a second, separately-shaped frozen PTv3 instance dedicated to the linear probe rather than a checkpoint substitution.

On licensing: both Sonata's and Utonia's weights are **CC-BY-NC 4.0**, non-commercial only. Neither can be a production candidate for a Thai government or commercial deployment as released — they are benchmark/research controls in the linear-probe comparison only, unless the team separately negotiates a commercial license for the underlying NC-restricted training data (HM3D, ArkitScenes, etc.). The code for both (Apache 2.0) carries no such restriction, so training an equivalent architecture from scratch on your own licensed data would sidestep the NC constraint — using the released weights directly does not.

**Comment by Chitchai-Jantanarak** [comment]
Resolved by the research comment above. Gist: both Sonata and Utonia weights are CC-BY-NC 4.0 (benchmark control only, not a production candidate); neither loads into this repo's PTv3 without a second, separately-shaped frozen encoder instance fed 9 channels (coord + color + normal, normals not computed anywhere in this repo today). Utonia first for the probe: it saw outdoor LiDAR (nuScenes, Waymo, SemanticKITTI); Sonata is indoor-only.
